From: IrfanMohammad <irfan@graphiant.com>
Date: Sat, 9 Mar 2024 21:35:57 +0000
Subject: bugfix shm_sub event type finished instead of none

For change_sub events, listeners should use SR_SUB_EV_FINISHED instead
of SR_SUB_EV_NONE to signal that they are done. Only the originators of
the event should set it to SR_SUB_EV_NONE.

sr_shmsub_notify_new_wrlock waits for SR_SUB_EV_NONE, and it can assume
that the previous event is finished but it's originator may not have
seen it yet.

To prevent this, previously orig_cid was used, but this is not reliable.
If a timeout happens while the listeners have the read lock,
the originator cannot clear orig_cid, and the shm becomes unusable
after this.

Since event is an atomic type, we can always set it even though we don't
have a WRITE lock.

RPCs will continue to use SR_SUB_EV_NONE for SR_SUB_EV_NOTIF as the
originators don't always wait, and setting event to SR_SUB_EV_FINISHED
is not needed.
---
 src/common.c       |  2 ++
 src/common_types.h |  1 +
 src/shm_sub.c      | 32 ++++++++++++++++++++------------
 src/shm_types.h    |  2 +-
 4 files changed, 24 insertions(+), 13 deletions(-)

diff --git a/src/common.c b/src/common.c
index f0dece3..5798047 100644
--- a/src/common.c
+++ b/src/common.c
@@ -4107,6 +4107,8 @@ sr_ev2str(sr_sub_event_t ev)
         return "success";
     case SR_SUB_EV_ERROR:
         return "error";
+    case SR_SUB_EV_FINISHED:
+        return "finished";
     case SR_SUB_EV_UPDATE:
         return "update";
     case SR_SUB_EV_CHANGE:
diff --git a/src/common_types.h b/src/common_types.h
index 48e02d5..0279d87 100644
--- a/src/common_types.h
+++ b/src/common_types.h
@@ -94,6 +94,7 @@ typedef enum {
     SR_SUB_EV_NONE = 0,         /**< No event. */
     SR_SUB_EV_SUCCESS,          /**< Event processed successfully by subscribers. */
     SR_SUB_EV_ERROR,            /**< Event failed to be processed by a subscriber. */
+    SR_SUB_EV_FINISHED,         /**< Event processing finished by subscribers. */
 
     SR_SUB_EV_UPDATE,           /**< New update event ready. */
     SR_SUB_EV_CHANGE,           /**< New change event ready. */
diff --git a/src/shm_sub.c b/src/shm_sub.c
index b199278..39859d7 100644
--- a/src/shm_sub.c
+++ b/src/shm_sub.c
@@ -346,7 +346,7 @@ sr_shmsub_notify_new_wrlock(sr_sub_shm_t *sub_shm, const char *shm_name, sr_sub_
         return err_info;
     }
 
-    if (sub_shm->orig_cid || (ATOMIC_LOAD_RELAXED(sub_shm->event) && (ATOMIC_LOAD_RELAXED(sub_shm->event) != lock_event))) {
+    if (ATOMIC_LOAD_RELAXED(sub_shm->event) && (ATOMIC_LOAD_RELAXED(sub_shm->event) != lock_event)) {
         /* instead of wating, try to recover the event immediately */
         sr_shmsub_recover(sub_shm);
     }
@@ -361,7 +361,7 @@ sr_shmsub_notify_new_wrlock(sr_sub_shm_t *sub_shm, const char *shm_name, sr_sub_
     /* wait until there is no event and there are no readers (just like write lock) */
     sr_timeouttime_get(&timeout_abs, SR_SUBSHM_LOCK_TIMEOUT);
     ret = 0;
-    while (!ret && (sub_shm->orig_cid || SR_RWLOCK_READERS(&sub_shm->lock) || (ATOMIC_LOAD_RELAXED(sub_shm->event) &&
+    while (!ret && (SR_RWLOCK_READERS(&sub_shm->lock) || (ATOMIC_LOAD_RELAXED(sub_shm->event) &&
             (ATOMIC_LOAD_RELAXED(sub_shm->event) != lock_event)))) {
         /* COND WAIT */
         ret = sr_cond_clockwait_retry(&sub_shm->lock.cond, &sub_shm->lock.mutex, &timeout_abs);
@@ -381,7 +381,7 @@ sr_shmsub_notify_new_wrlock(sr_sub_shm_t *sub_shm, const char *shm_name, sr_sub_
 
         if (ret == ETIMEDOUT) {
             /* try to recover the event again in case the originator crashed later */
-            if (last_event || sub_shm->orig_cid) {
+            if (last_event) {
                 sr_shmsub_recover(sub_shm);
                 last_event = ATOMIC_LOAD_RELAXED(sub_shm->event);
             }
@@ -423,7 +423,7 @@ sr_shmsub_notify_new_wrlock(sr_sub_shm_t *sub_shm, const char *shm_name, sr_sub_
 
 event_handled:
     /* we have write lock and the expected event */
-    assert(!sub_shm->orig_cid);
+    sub_shm->orig_cid = 0;
     return NULL;
 }
 
@@ -437,6 +437,7 @@ event_handled:
  * @param[in] request_id Current request ID of the event in @p sub_shm.
  * @param[in] expected_ev Expected event. Can be:
  *              ::SR_SUB_EV_NONE - just wait until the event is processed, SHM will not be accessed,
+ *              ::SR_SUB_EV_FINISHED - just wait until the event is processed, SHM will not be accessed,
  *              ::SR_SUB_EV_SUCCESS - an answer (success/error) is expected but SHM will not be accessed, so
  *                                    success (never error) event is cleared,
  *              ::SR_SUB_EV_ERROR - an answer is expected and SHM will be further accessed so do not clear any events.
@@ -461,7 +462,7 @@ _sr_shmsub_notify_wait_wr(sr_sub_shm_t *sub_shm, sr_sub_event_t event, uint32_t
     int ret, write_lock = 0;
     char buf[64] = "";
 
-    assert((expected_ev == SR_SUB_EV_NONE) || (expected_ev == SR_SUB_EV_SUCCESS) || (expected_ev == SR_SUB_EV_ERROR));
+    assert((expected_ev == SR_SUB_EV_NONE) || SR_IS_NOTIFY_EVENT(expected_ev));
     assert(shm_data_sub->fd > -1);
 
     *lock_lost = 0;
@@ -597,8 +598,10 @@ event_handled:
             return err_info;
         }
     } else {
-        /* we expect no event */
-        if (sub_shm->event != SR_SUB_EV_NONE) {
+        /* we expect a finished or none event */
+        if (ATOMIC_LOAD_RELAXED(sub_shm->event) == expected_ev) {
+            ATOMIC_STORE_RELAXED(sub_shm->event, SR_SUB_EV_NONE);
+        } else {
             sr_errinfo_new(&err_info, SR_ERR_INTERNAL, "Unexpected sub SHM event \"%s\" (expected \"%s\").",
                     sr_ev2str(last_event), sr_ev2str(expected_ev));
             return err_info;
@@ -616,6 +619,7 @@ event_handled:
  * @param[in] sub_shm Subscription SHM.
  * @param[in] expected_ev  Expected event. Can be:
  *              ::SR_SUB_EV_NONE - just wait until the event is processed, SHM will not be accessed,
+ *              ::SR_SUB_EV_FINISHED - just wait until the event is processed, SHM will not be accessed,
  *              ::SR_SUB_EV_SUCCESS - an answer (success/error) is expected but SHM will not be accessed, so
  *                                    success (never error) event is cleared,
  *              ::SR_SUB_EV_ERROR - an answer is expected and SHM will be further accessed so do not clear any events.
@@ -655,6 +659,7 @@ sr_shmsub_notify_wait_wr(sr_sub_shm_t *sub_shm, sr_sub_event_t expected_ev, int
  * @param[in] notify_count Size of the array.
  * @param[in] expected_ev  Expected event. Can be:
  *              ::SR_SUB_EV_NONE - just wait until the event is processed, SHM will not be accessed,
+ *              ::SR_SUB_EV_FINISHED - just wait until the event is processed, SHM will not be accessed,
  *              ::SR_SUB_EV_SUCCESS - an answer (success/error) is expected but SHM will not be accessed, so
  *                                    success (never error) event is cleared,
  *              ::SR_SUB_EV_ERROR - an answer is expected and SHM will be further accessed so do not clear any events.
@@ -2055,7 +2060,7 @@ sr_shmsub_change_notify_change_done(struct sr_mod_info_s *mod_info, const char *
 
         /* wait until the events are processed */
         if ((err_info = sr_shmsub_notify_many_wait_wr((struct sr_shmsub_many_info_s *)notify_subs, sizeof *notify_subs,
-                notify_count, SR_SUB_EV_NONE, 1, cid, timeout_ms))) {
+                notify_count, SR_SUB_EV_FINISHED, 1, cid, timeout_ms))) {
             goto cleanup;
         }
 
@@ -2275,7 +2280,7 @@ sr_shmsub_change_notify_change_abort(struct sr_mod_info_s *mod_info, const char
 
         /* wait until the events are processed */
         if ((err_info = sr_shmsub_notify_many_wait_wr((struct sr_shmsub_many_info_s *)notify_subs, sizeof *notify_subs,
-                notify_count, SR_SUB_EV_NONE, 1, cid, timeout_ms))) {
+                notify_count, SR_SUB_EV_FINISHED, 1, cid, timeout_ms))) {
             goto cleanup;
         }
 
@@ -3061,7 +3066,7 @@ clear_shm:
         }
 
         /* wait until the event is processed */
-        if ((err_info = sr_shmsub_notify_wait_wr((sr_sub_shm_t *)multi_sub_shm, SR_SUB_EV_NONE, 1, conn->cid,
+        if ((err_info = sr_shmsub_notify_wait_wr((sr_sub_shm_t *)multi_sub_shm, SR_SUB_EV_FINISHED, 1, conn->cid,
                 &shm_data_sub, timeout_ms, &lock_lost, &cb_err_info))) {
             if (lock_lost) {
                 goto cleanup;
@@ -3314,8 +3319,11 @@ sr_shmsub_multi_listen_write_event(sr_multi_sub_shm_t *multi_sub_shm, uint32_t v
             break;
         case SR_SUB_EV_DONE:
         case SR_SUB_EV_ABORT:
-        case SR_SUB_EV_NOTIF:
             /* notifier does not wait for these events */
+            assert(!err_code);
+            ATOMIC_STORE_RELAXED(multi_sub_shm->event, SR_SUB_EV_FINISHED);
+            break;
+        case SR_SUB_EV_NOTIF:
             assert(!err_code);
             ATOMIC_STORE_RELAXED(multi_sub_shm->event, SR_SUB_EV_NONE);
             break;
@@ -3578,7 +3586,7 @@ sr_shmsub_change_listen_process_module_events(struct modsub_change_s *change_sub
         /* no new module event */
         goto cleanup;
     }
-    
+
     first_new_event_sub_idx = i;
 
     /* SUB READ LOCK */
diff --git a/src/shm_types.h b/src/shm_types.h
index 2b31101..86f735a 100644
--- a/src/shm_types.h
+++ b/src/shm_types.h
@@ -41,7 +41,7 @@
         || (ev == SR_SUB_EV_NOTIF))
 
 /** Whether an event is one to be processed by the originators. */
-#define SR_IS_NOTIFY_EVENT(ev) ((ev == SR_SUB_EV_SUCCESS) || (ev == SR_SUB_EV_ERROR))
+#define SR_IS_NOTIFY_EVENT(ev) ((ev == SR_SUB_EV_SUCCESS) || (ev == SR_SUB_EV_ERROR) || (ev == SR_SUB_EV_FINISHED))
 
 /**
  * @brief Mod SHM dependency type.
